File: app/src/main/java/tpcreative/co/qrscanner/ui/history/HistoryPresenter.java
package tpcreative.co.qrscanner.ui.history;
import com.google.gson.Gson;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import tpcreative.co.qrscanner.common.Utils;
import tpcreative.co.qrscanner.common.presenter.Presenter;
import tpcreative.co.qrscanner.helper.SQLiteHelper;
import tpcreative.co.qrscanner.model.HistoryModel;
import tpcreative.co.qrscanner.model.TypeCategories;

public class HistoryPresenter extends Presenter<HistoryView> {

    protected List<TypeCategories> mListCategories;
    protected List<HistoryModel> mList;
    protected HistoryModel mLatestValue;
    private int i = 0;
    private static final String TAG = HistoryPresenter.class.getSimpleName();

    public HistoryPresenter(){
        mListCategories = new ArrayList<>();
        mList = new ArrayList<>();
        mLatestValue = new HistoryModel();
    }

    public List<HistoryModel>getLatestList(List<HistoryModel> mData){
        List<HistoryModel> mList = new ArrayList<>();
        if (mData.size()>0){
            mLatestValue = mData.get(0);
            for (HistoryModel index : mData){
                if (index.createType.equals(mLatestValue.createType)){
                    index.typeCategories = new TypeCategories(0,mLatestValue.createType);
                    mList.add(index);
                }
            }
        }
        Collections.sort(mList, new Comparator<HistoryModel>() {
            public int compare(HistoryModel o1, HistoryModel o2) {
                return (Utils.getMilliseconds(o1.updatedDateTime)<Utils.getMilliseconds(o2.updatedDateTime)) ? 0 : 1;
            }
        });
        return mList;
    }

    public Map<String, HistoryModel> getUniqueList(){
        HistoryView view = view();
        final List<HistoryModel> histories = SQLiteHelper.getHistoryList();
        Utils.Log(TAG,"History list "+ new Gson().toJson(histories));
        if (histories==null){
            return new HashMap<>();
        }
        Utils.Log(TAG,new Gson().toJson(histories));
        Map<String, HistoryModel> hashMap = new HashMap<>();
        for (HistoryModel index : histories){
            hashMap.put(index.createType,index);
        }
        mListCategories.clear();
        i=1;
        for (Map.Entry<String, HistoryModel>map : hashMap.entrySet()){
            mListCategories.add(new TypeCategories(i,map.getKey()));
            i+=1;
        }
        return hashMap;
    }

    public List<HistoryModel> getListGroup(){
        getUniqueList();
        HistoryView view = view();
        final List<HistoryModel> list = SQLiteHelper.getHistoryList();
        List<HistoryModel> mList = new ArrayList<>();
        final List<HistoryModel> mLatestType = getLatestList(list);
        Utils.Log(TAG,"Latest list "+ new Gson().toJson(mLatestType));
        Utils.Log(TAG,"Latest object " + new Gson().toJson(mLatestValue));
        for (TypeCategories index : mListCategories){
            for (HistoryModel history : list){
                if (index.type.equals(history.createType) && !index.type.equals(mLatestValue.createType)){
                    final HistoryModel result = history;
                    result.typeCategories = index;
                    mList.add(result);
                }
            }
        }

        /*Added latest list to ArrayList*/
        mLatestType.addAll(mList);
        this.mList.clear();
        this.mList.addAll(mLatestType);
        return mLatestType;
    }

    public int getCheckedCount(){
        int count = 0;
        for (HistoryModel index : mList){
            if (index.isChecked()){
                count+=1;
            }
        }
        return count;
    }

    public void deleteItem(){
        HistoryView view = view();
        final List<HistoryModel> list = mList;
        for (HistoryModel index : list){
            if (index.isDeleted()){
                if (index.isChecked()){
                    SQLiteHelper.onDelete(index);
                }
            }
        }
        getListGroup();
        view.updateView();
    }

}

File: app/src/main/java/tpcreative/co/qrscanner/ui/history/HistoryPresenter.kt
package tpcreative.co.qrscanner.ui.history

import com.google.gson.Gson
import tpcreative.co.qrscanner.common.*
import tpcreative.co.qrscanner.common.presenter.Presenter
import tpcreative.co.qrscanner.helper.SQLiteHelper
import tpcreative.co.qrscanner.model.HistoryModel
import tpcreative.co.qrscanner.model.TypeCategories
import java.util.*

class HistoryPresenter : Presenter<HistoryView?>() {
    protected var mListCategories: MutableList<TypeCategories?>?
    var mList: MutableList<HistoryModel?>?
    protected var mLatestValue: HistoryModel?
    private var i = 0
    fun getLatestList(mData: MutableList<HistoryModel?>?): MutableList<HistoryModel?>? {
        val mList: MutableList<HistoryModel?> = ArrayList()
        if (mData.size > 0) {
            mLatestValue = mData.get(0)
            for (index in mData) {
                if (index.createType == mLatestValue.createType) {
                    index.typeCategories = TypeCategories(0, mLatestValue.createType)
                    mList.add(index)
                }
            }
        }
        Collections.sort(mList) { o1, o2 -> if (Utils.getMilliseconds(o1.updatedDateTime) < Utils.getMilliseconds(o2.updatedDateTime)) 0 else 1 }
        return mList
    }

    fun getUniqueList(): MutableMap<String?, HistoryModel?>? {
        val view = view()
        val histories = SQLiteHelper.getHistoryList()
        Utils.Log(TAG, "History list " + Gson().toJson(histories))
        if (histories == null) {
            return HashMap()
        }
        Utils.Log(TAG, Gson().toJson(histories))
        val hashMap: MutableMap<String?, HistoryModel?> = HashMap()
        for (index in histories) {
            hashMap[index.createType] = index
        }
        mListCategories.clear()
        i = 1
        for ((key) in hashMap) {
            mListCategories.add(TypeCategories(i, key))
            i += 1
        }
        return hashMap
    }

    fun getListGroup(): MutableList<HistoryModel?>? {
        getUniqueList()
        val view = view()
        val list = SQLiteHelper.getHistoryList()
        val mList: MutableList<HistoryModel?> = ArrayList()
        val mLatestType = getLatestList(list)
        Utils.Log(TAG, "Latest list " + Gson().toJson(mLatestType))
        Utils.Log(TAG, "Latest object " + Gson().toJson(mLatestValue))
        for (index in mListCategories) {
            for (history in list) {
                if (index.type == history.createType && index.type != mLatestValue.createType) {
                    history.typeCategories = index
                    mList.add(history)
                }
            }
        }

        /*Added latest list to ArrayList*/mLatestType.addAll(mList)
        this.mList.clear()
        this.mList.addAll(mLatestType)
        return mLatestType
    }

    fun getCheckedCount(): Int {
        var count = 0
        for (index in mList) {
            if (index.isChecked()) {
                count += 1
            }
        }
        return count
    }

    fun deleteItem() {
        val view = view()
        val list = mList
        for (index in list) {
            if (index.isDeleted()) {
                if (index.isChecked()) {
                    SQLiteHelper.onDelete(index)
                }
            }
        }
        getListGroup()
        view.updateView()
    }

    companion object {
        private val TAG = HistoryPresenter::class.java.simpleName
    }

    init {
        mListCategories = ArrayList()
        mList = ArrayList()
        mLatestValue = HistoryModel()
    }
}